testament/testament.nim

```diff
@@ -2017,6 +2080,17 @@ proc runTests(execState: var Execution) =
   startTimes = newSeq[float](batchSize)
   endTimes = newSeq[float](batchSize)
 
+  # REFACTOR: this function requires additional context to be available,
```

Comment:
making the callbacks top level module procs also doesn't help as it's not really part of the public surface area and only ever used here.

Reply:
I'm thinking about using

```nim
proc createBatchExecutionTriggers(batchStore: BatchStore, batchId: BatchId): tuple[onStart: proc(id: int), onProcess, onComplete: proc(id: int, p: Process)] = 
  result.onProcess = proc(id: int) = 
    batchStore[batchId].startTimes[id] = epochTime()
    # ... etc
```